### PR TITLE
Bugfix/azure wait for disk detach

### LINKED ISSUE #242: Azure machines being stuck on deletion
## Issue
It is often observed that Azure machines are stuck on deletion

## Possible cause
As suggested by colleagues at Microsoft, it is observed that detaching of PVs could take longer than 5mins leading to this in a consistent state on hung VMs

## Solution
Need to enhance drain timeouts?

### PR DESCRIPTION
**Pull request by prashanth26**
**What this PR does / why we need it**:
This PR attempts to fix the Azure issue of VMs stuck in deletion state. Now Azure waits for disks being detached and drains to complete before deleting the VM.

**Which issue(s) this PR fixes**:
Fixes #242 

**Special notes for your reviewer**:

**Release note**:

```noteworthy user
The drain is always invoked even the case of forceful deletion
```
```noteworthy user
Drain now tries to evict pods and if eviction fails, it forcefully deletes the pods
```
```improvement operator
Azure explicitly detaches data disks before VM deletion
```

### CODE REVIEW

**Comment by prashanth26** [comment]
@hardikdr /needs-review

File: pkg/controller/machine.go [hunk 1]
```diff
@@ -470,7 +468,7 @@ func (c *controller) machineDelete(machine *v1alpha1.Machine, driver driver.Driv
 					c.targetCoreClient,
 					timeOutDuration, // TODO: Will need to configure timeout
 					nodeName,
-					-1,
+					int(timeOutDuration.Seconds()),
```
Comment (hardikdr) [inline pkg/controller/machine.go]: What is it for ?

File: pkg/driver/driver_azure.go [hunk 2]
```diff
@@ -380,6 +379,33 @@ func (d *AzureDriver) getvms(machineID string, listOfVMs VMs) error {
...
+		glog.Errorf("Failed to list VMs. Error Message - %s", err)
+		return err
+	}
+	metrics.APIRequestCount.With(prometheus.Labels{"provider": "azure", "service": "virtual_machine"}).Inc()
+
+	if len(*vm.StorageProfile.DataDisks) > 0 {
+		// There are disks attached hence need to detach them
+		vm.StorageProfile.DataDisks = &[]compute.DataDisk{}
+
+		_, errChan := vmClient.CreateOrUpdate(d.AzureMachineClass.Spec.ResourceGroup, machineID, vm, cancel)
```
Comment (hardikdr) [inline pkg/driver/driver_azure.go]: Can you please explain how does it work? 
Where are we making explicit detach calls to VM? , or does the empty-disk array instructs API to delete all disk?

File: pkg/controller/machine.go (hunk 1 above)
Comment (prashanth26) [reply pkg/controller/machine.go]: I thought to cap the graceful termination of any pod by the drain timeout should help in overall drain option going through in the drain timeout. But i dunno if it works as expected.

I am hoping that if a pod is set to a graceful termination of `2Hours`, and our drain timeout is '5mins` we max would only give 5mins?

File: pkg/driver/driver_azure.go (hunk 2 above)
Comment (prashanth26) [reply pkg/driver/driver_azure.go]: Yes hardik. VM with an update call with empty data disks is how detachment of data disks is done in Azure. They lack a proper SDK call for it. Refer here - https://github.com/Azure/azure-sdk-for-go/issues/1638#issuecomment-384733952

File: pkg/controller/machine.go (hunk 1 above)
Comment (prashanth26) [reply pkg/controller/machine.go]: I have reverted this change here - https://github.com/gardener/machine-controller-manager/pull/248/commits/7510f9ce74835b969614d94cb0cfa65f392becea#diff-d8287fe74b5273163c9f6b6c635ad912R475.

Comment (hardikdr) [reply pkg/controller/machine.go]: Thanks.

**Pull request merged**